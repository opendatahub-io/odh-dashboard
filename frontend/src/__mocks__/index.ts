export * from './mockDashboardConfig';
export * from './mockDataSciencePipelinesApplicationK8sResource';
export * from './mockK8sResourceList';
export * from './mockProjectK8sResource';
export * from './mockRouteK8sResource';
export * from './mockStatus';
export * from './mockRunKF';
export * from './mockRecurringRunKF';
export * from './mockExperimentKF';
export * from './mockPipelineVersionsProxy';
export * from './mockPipelinesProxy';
export * from './mockDscStatus';
export * from './mockNotebookK8sResource';
export * from './mockNotebookState';
export * from './mockPipelineKF';
export * from './mockSecretK8sResource';
export * from './mockGoogleRpcStatusKF';
export * from './mockK8sStatus';
export * from './mockComponents';
export * from './mockRegisteredModel';
export * from './mockModelVersion';
export * from './mockModelVersionList';
export * from './mockModelArtifactList';
export * from './mockModelRegistryService';
export * from './mockServingRuntimeK8sResource';
export * from './mockInferenceServiceK8sResource';
export * from './mockStorageClasses';
export * from './mockConfigMapSecrets';
