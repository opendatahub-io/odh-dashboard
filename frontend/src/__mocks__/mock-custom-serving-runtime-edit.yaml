apiVersion: serving.kserve.io/v1alpha1
kind: ServingRuntime
metadata:
  name: template-1
  annotations:
    openshift.io/display-name: Updated Caikit
  labels:
    opendatahub.io/dashboard: 'true'
spec:
  builtInAdapter:
    memBufferBytes: 134217728
    modelLoadingTimeoutMillis: 90000
    runtimeManagementPort: 8888
    serverType: ovms
  containers:
    - args:
        - --port=8001
        - --rest_port=8888
        - --config_path=/models/model_config_list.json
        - --file_system_poll_wait_seconds=0
        - --grpc_bind_address=127.0.0.1
        - --rest_bind_address=127.0.0.1
        - --target_device=NVIDIA
      image: quay.io/modh/openvino-model-server@sha256:c89f76386bc8b59f0748cf173868e5beef21ac7d2f78dada69089c4d37c44116
      name: ovms
      resources:
        limits:
          cpu: '0'
          memory: 0Gi
        requests:
          cpu: '0'
          memory: 0Gi
  grpcDataEndpoint: port:8001
  grpcEndpoint: port:8085
  multiModel: true
  protocolVersions:
    - grpc-v1
  supportedModelFormats:
    - autoSelect: true
      name: openvino_ir
      version: opset1
    - autoSelect: true
      name: onnx
      version: '1'
