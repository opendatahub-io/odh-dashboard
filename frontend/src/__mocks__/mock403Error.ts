import { K8sStatus } from '~/k8sTypes';

export const mock403Error = ({ message = '403 Forbidden' }: Partial<K8sStatus>): K8sStatus => ({
  kind: 'Status',
  apiVersion: 'v1',
  status: 'Failure',
  message,
  reason: 'Forbidden',
  code: 403,
});
