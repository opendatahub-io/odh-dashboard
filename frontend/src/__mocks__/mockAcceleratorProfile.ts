import { AcceleratorProfileKind } from '~/k8sTypes';
import { Toleration, TolerationEffect, TolerationOperator } from '~/types';
import { genUID } from './mockUtils';

type MockResourceConfigType = {
  name?: string;
  namespace?: string;
  uid?: string;
  displayName?: string;
  identifier?: string;
  description?: string;
  enabled?: boolean;
  tolerations?: Toleration[];
};

export const mockAcceleratorProfile = ({
  name = 'migrated-gpu',
  namespace = 'test-project',
  uid = genUID('service'),
  displayName = 'Nvidia GPU',
  identifier = 'nvidia.com/gpu',
  description = '',
  enabled = true,
  tolerations = [
    {
      key: 'nvidia.com/gpu',
      operator: TolerationOperator.EXISTS,
      effect: TolerationEffect.NO_SCHEDULE,
    },
  ],
}: MockResourceConfigType): AcceleratorProfileKind => ({
  apiVersion: 'dashboard.opendatahub.io/v1',
  kind: 'AcceleratorProfile',
  metadata: {
    creationTimestamp: '2023-03-17T16:12:41Z',
    generation: 1,
    name,
    namespace,
    resourceVersion: '1309350',
    uid,
  },
  spec: {
    identifier,
    displayName,
    enabled,
    tolerations,
    description,
  },
});

export const mockGlobalScopedAcceleratorProfiles = [
  mockAcceleratorProfile({
    name: 'small-profile-global',
    displayName: 'Small Profile Global',
    namespace: 'opendatahub',
    tolerations: [
      {
        effect: TolerationEffect.NO_SCHEDULE,
        key: 'NotebooksOnlyChange',
        operator: TolerationOperator.EXISTS,
      },
    ],
  }),
  mockAcceleratorProfile({
    name: 'large-profile-global',
    displayName: 'Large Profile Global',
    namespace: 'opendatahub',
  }),
];

export const mockProjectScopedAcceleratorProfiles = [
  mockAcceleratorProfile({
    name: 'small-profile',
    displayName: 'Small Profile',
    namespace: 'test-project',
    tolerations: [
      {
        effect: TolerationEffect.NO_SCHEDULE,
        key: 'NotebooksOnlyChange',
        operator: TolerationOperator.EXISTS,
      },
    ],
  }),
  mockAcceleratorProfile({
    name: 'large-profile-1',
    displayName: 'Large Profile-1',
    namespace: 'test-project',
  }),
];
