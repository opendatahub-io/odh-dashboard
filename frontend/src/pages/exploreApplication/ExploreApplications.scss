.odh-explore-apps {
  &__body {
    display: flex;
    flex-direction: column;
  }
}
