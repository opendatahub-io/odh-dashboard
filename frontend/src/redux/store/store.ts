import { createStore, applyMiddleware, compose } from 'redux';
import reduxThunk from 'redux-thunk';
import rootReducer from '../reducers/rootReducers';

// eslint-disable-next-line
const composeEnhancers = (window as any).__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;

const configureStore = () => {
  return createStore(rootReducer, composeEnhancers(applyMiddleware(reduxThunk)));
};

export const store = configureStore();
