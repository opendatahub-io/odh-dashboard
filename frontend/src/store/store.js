import { createStore, applyMiddleware, compose } from "redux";
import reduxThunk from "redux-thunk";
import rootReducer from "./reducers";

const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;

const configureStore = () => {
  return createStore(rootReducer, composeEnhancers(applyMiddleware(reduxThunk)));
};

export const store = configureStore();
