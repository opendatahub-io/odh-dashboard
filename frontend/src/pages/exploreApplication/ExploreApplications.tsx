import * as React from 'react';
import { useNavigate } from 'react-router-dom';
import classNames from 'classnames';
import * as _ from 'lodash';
import {
  Drawer,
  DrawerContent,
  DrawerContentBody,
  Gallery,
  PageSection,
} from '@patternfly/react-core';
import { useWatchComponents } from '../../utilities/useWatchComponents';
import OdhExploreCard from '../../components/OdhExploreCard';
import ApplicationsPage from '../ApplicationsPage';
import { OdhApplication } from '../../types';
import GetStartedPanel from './GetStartedPanel';
import { useQueryParams } from '../../utilities/useQueryParams';
import { removeQueryArgument, setQueryArgument } from '../../utilities/router';
import { fireTrackingEvent } from '../../utilities/segmentIOUtils';
import { ODH_PRODUCT_NAME } from '../../utilities/const';
import { useAppContext } from '../../app/AppContext';

import './ExploreApplications.scss';

const description = `Add optional applications to your ${ODH_PRODUCT_NAME} instance.`;
const disabledDescription = `View optional applications for your ${ODH_PRODUCT_NAME} instance. Contact an administrator to install these applications.`;

type ExploreApplicationsInnerProps = {
  loaded: boolean;
  isEmpty: boolean;
  loadError?: Error;
  exploreComponents: OdhApplication[];
  selectedComponent?: OdhApplication;
  updateSelection: (selectedId?: string | null) => void;
};

const ExploreApplicationsInner: React.FC<ExploreApplicationsInnerProps> = React.memo(
  ({ loaded, isEmpty, loadError, exploreComponents, selectedComponent, updateSelection }) => {
    const { dashboardConfig } = useAppContext();
    const disableInfo = dashboardConfig.spec.dashboardConfig.disableInfo;
    const bodyClasses = classNames('odh-explore-apps__body', {
      'm-side-panel-open': !!selectedComponent,
    });
    const [enableApp, setEnableApp] = React.useState<OdhApplication>();

    return (
      <Drawer
        data-id="explore-applications"
        isExpanded={!disableInfo && !!selectedComponent}
        isInline
      >
        <DrawerContent
          panelContent={
            <GetStartedPanel
              onClose={() => updateSelection()}
              selectedApp={selectedComponent}
              onEnable={() => setEnableApp(selectedComponent)}
            />
          }
        >
          <DrawerContentBody className={bodyClasses}>
            <ApplicationsPage
              title="Explore"
              description={disableInfo ? disabledDescription : description}
              loaded={loaded}
              empty={isEmpty}
              loadError={loadError}
            >
              {!isEmpty ? (
                <div className="odh-dashboard__page-content" data-id="page-content">
                  <PageSection>
                    <Gallery role="list" className="odh-explore-apps__gallery" hasGutter>
                      {exploreComponents.map((c) => (
                        <OdhExploreCard
                          key={c.metadata.name}
                          odhApp={c}
                          isSelected={selectedComponent?.metadata.name === c.metadata.name}
                          onSelect={() => {
                            updateSelection(c.metadata.name);
                            fireTrackingEvent('Explore card clicked', {
                              name: c.metadata.name,
                            });
                          }}
                          disableInfo={disableInfo}
                          enableOpen={c.metadata.name === enableApp?.metadata.name}
                          onEnableClose={() => setEnableApp(undefined)}
                        />
                      ))}
                    </Gallery>
                  </PageSection>
                </div>
              ) : null}
            </ApplicationsPage>
          </DrawerContentBody>
        </DrawerContent>
      </Drawer>
    );
  },
);
ExploreApplicationsInner.displayName = 'ExploreApplicationsInner';

const ExploreApplications: React.FC = () => {
  const { components, loaded, loadError } = useWatchComponents(false);
  const navigate = useNavigate();
  const queryParams = useQueryParams();
  const selectedId = queryParams.get('selectId');
  const [selectedComponent, setSelectedComponent] = React.useState<OdhApplication>();
  const isEmpty = !components || components.length === 0;

  const updateSelection = React.useCallback(
    (selectedId?: string | null): void => {
      const selection = components.find((c) => c.metadata.name && c.metadata.name === selectedId);
      if (selectedId && selection) {
        setQueryArgument(navigate, 'selectId', selectedId);
        setSelectedComponent(selection);
        return;
      }

      setSelectedComponent(undefined);
      removeQueryArgument(navigate, 'selectId');
    },
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [components],
  );

  const exploreComponents = React.useMemo<OdhApplication[]>(() => {
    return _.cloneDeep(components).sort((a, b) =>
      a.spec.displayName.localeCompare(b.spec.displayName),
    );
  }, [components]);

  React.useEffect(() => {
    if (components?.length > 0) {
      updateSelection(selectedId);
    }
  }, [updateSelection, selectedId, components]);

  return (
    <ExploreApplicationsInner
      loaded={loaded}
      isEmpty={isEmpty}
      loadError={loadError}
      exploreComponents={exploreComponents}
      selectedComponent={selectedComponent}
      updateSelection={updateSelection}
    />
  );
};

export default ExploreApplications;
