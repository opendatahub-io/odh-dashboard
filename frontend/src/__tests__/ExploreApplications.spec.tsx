import * as React from 'react';
import { act } from 'react-dom/test-utils';
import { store } from '../redux/store/store';
import { Provider } from 'react-redux';
import { BrowserRouter as Router } from 'react-router-dom';
import { mount } from 'enzyme';
import ExploreApplications from '../pages/exploreApplication/ExploreApplications';
import { mockExploreApplications } from '../../__mocks__/mockExploreApplications';
import { mockGettingStartedDoc } from '../../__mocks__/mockGettingStartedDoc';

jest.mock('react-router-dom', () => {
  const reactRouterDom = jest.requireActual('react-router-dom');
  return {
    ...reactRouterDom,
    useLocation: () => '/',
  };
});
jest.mock('react', () => {
  const React = jest.requireActual('react');
  return {
    ...React,
    Suspense: ({ children }) => children,
    lazy: (factory) => factory(),
  };
});
jest.mock('../utilities/useWatchComponents', () => ({
  useWatchComponents: () => ({
    loaded: true,
    loadError: null,
    components: mockExploreApplications,
  }),
}));
jest.mock('../utilities/useGettingStarted', () => ({
  useGettingStarted: () => ({
    odhGettingStarted: mockGettingStartedDoc,
    loaded: true,
    loadError: null,
  }),
}));
jest.mock('../utilities/router', () => ({
  setQueryArgument: () => {
    return;
  },
  removeQueryArgument: () => {
    return;
  },
}));

describe('ExploreApplications', () => {
  it('should display available applications', () => {
    const wrapper = mount(<ExploreApplications />);

    expect(wrapper.find('.odh-explore-apps__body').exists()).toBe(true);
    const cards = wrapper.find('.pf-m-selectable.odh-card');
    expect(cards.length).toBe(2);

    expect(wrapper.html()).toMatchSnapshot();
    wrapper.unmount();
  });

  it('should show the getting started panel on card click', () => {
    const wrapper = mount(<ExploreApplications />);
    expect(wrapper.find('.m-side-panel-open').exists()).toBe(false);

    act(() => {
      const card = wrapper.find('.pf-m-selectable.odh-card').at(0);
      card.simulate('click');
    });
    wrapper.update();

    expect(wrapper.find('.m-side-panel-open').exists()).toBe(true);
    expect(wrapper.html()).toMatchSnapshot();

    act(() => {
      const close = wrapper.find('.pf-c-drawer__close .pf-c-button');
      close.simulate('click');
    });
    wrapper.update();
    expect(wrapper.find('.m-side-panel-open').exists()).toBe(false);
    wrapper.unmount();
  });

  it('should show the enable modal when clicked', () => {
    const wrapper = mount(
      <Provider store={store}>
        <Router>
          <ExploreApplications />
        </Router>
      </Provider>,
    );

    // First app is enabled, there should be no enable button
    act(() => {
      const card = wrapper.find('.pf-m-selectable.odh-card').at(0);
      card.simulate('click');
    });
    wrapper.update();
    expect(wrapper.find('.m-side-panel-open').exists()).toBe(true);
    expect(
      wrapper.find('.odh-get-started__button-panel .pf-c-button.pf-m-secondary').exists(),
    ).toBe(false);

    // Second app is enable-able, there should be an enable button
    act(() => {
      const card = wrapper.find('.pf-m-selectable.odh-card').at(1);
      card.simulate('click');
    });
    wrapper.update();
    expect(wrapper.find('.m-side-panel-open').exists()).toBe(true);
    expect(
      wrapper.find('.odh-get-started__button-panel .pf-c-button.pf-m-secondary').exists(),
    ).toBe(true);

    expect(wrapper.find('.odh-enable-modal').exists()).toBe(false);
    act(() => {
      const close = wrapper.find('.odh-get-started__button-panel .pf-c-button.pf-m-secondary');
      close.simulate('click');
    });
    wrapper.update();
    expect(wrapper.find('.odh-enable-modal').exists()).toBe(true);

    expect(wrapper.find('.pf-c-modal-box.odh-enable-modal').html()).toMatchSnapshot();

    act(() => {
      const close = wrapper.find(
        '.odh-enable-modal .pf-c-modal-box__footer .pf-c-button.pf-m-link',
      );
      close.simulate('click');
    });
    wrapper.update();
    expect(wrapper.find('.odh-enable-modal').exists()).toBe(false);

    wrapper.unmount();
  });
});
