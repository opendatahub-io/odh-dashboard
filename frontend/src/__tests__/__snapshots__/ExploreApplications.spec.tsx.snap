// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExploreApplications should display available applications 1`] = `"<section class=\\"pf-c-page__main-section pf-m-light odh-apps__heading\\"><div class=\\"pf-c-content odh-apps__heading__text\\"><h1 data-pf-content=\\"true\\" class=\\"\\">Explore</h1><p data-pf-content=\\"true\\" class=\\"\\">Add optional applications to your Open Data Hub instance.</p></div></section><div class=\\"pf-c-drawer pf-m-inline\\"><div class=\\"pf-c-drawer__main\\"><div class=\\"pf-c-drawer__content\\"><div class=\\"pf-c-drawer__body odh-explore-apps__body\\"><div class=\\"odh-dashboard__page-content\\"><section class=\\"pf-c-page__main-section\\"><div class=\\"pf-l-gallery pf-m-gutter odh-explore-apps__gallery\\"><article id=\\"jupyterhub\\" class=\\"pf-c-card pf-m-hoverable pf-m-selectable odh-card\\" tabindex=\\"0\\" data-ouia-component-type=\\"PF4/Card\\" data-ouia-safe=\\"true\\" data-ouia-component-id=\\"OUIA-Generated-Card-1\\"><div class=\\"pf-c-card__header\\"><img class=\\"pf-c-brand odh-card__header-brand odh-card__header-brand\\" src=\\"images/jupyterhub.svg\\" alt=\\"JupyterHub\\"></div><div class=\\"pf-c-card__title\\"><span class=\\"odh-card__title-supported\\">JupyterHub</span><div><span class=\\"odh-card__provider\\">by Jupyter</span></div></div><div class=\\"pf-c-card__body\\">A multi-user version of the notebook designed for companies, classrooms and research labs.</div></article><article id=\\"test-app\\" class=\\"pf-c-card pf-m-hoverable pf-m-selectable odh-card\\" tabindex=\\"0\\" data-ouia-component-type=\\"PF4/Card\\" data-ouia-safe=\\"true\\" data-ouia-component-id=\\"OUIA-Generated-Card-2\\"><div class=\\"pf-c-card__header\\"><img class=\\"pf-c-brand odh-card__header-brand odh-card__header-brand\\" src=\\"images/jupyterhub.svg\\" alt=\\"Test App\\"></div><div class=\\"pf-c-card__title\\"><span class=\\"odh-card__title-supported\\">Test App</span><div><span class=\\"odh-card__provider\\">by Test</span></div></div><div class=\\"pf-c-card__body\\">Some description</div></article></div></section></div></div></div></div></div>"`;

exports[`ExploreApplications should show the enable modal when clicked 1`] = `"<div id=\\"pf-modal-part-0\\" data-ouia-component-type=\\"PF4/ModalContent\\" data-ouia-safe=\\"true\\" data-ouia-component-id=\\"OUIA-Generated-Modal-small-1\\" role=\\"dialog\\" aria-label=\\"Enable Test Enable\\" aria-labelledby=\\"pf-modal-part-0 pf-modal-part-1\\" aria-describedby=\\"pf-modal-part-2\\" aria-modal=\\"true\\" class=\\"pf-c-modal-box odh-enable-modal pf-m-sm\\"><button aria-disabled=\\"false\\" aria-label=\\"Close\\" class=\\"pf-c-button pf-m-plain\\" type=\\"button\\" data-ouia-component-type=\\"PF4/Button\\" data-ouia-safe=\\"true\\" data-ouia-component-id=\\"OUIA-Generated-Button-plain-3\\"><svg style=\\"vertical-align: -0.125em;\\" fill=\\"currentColor\\" height=\\"1em\\" width=\\"1em\\" viewBox=\\"0 0 352 512\\" aria-hidden=\\"true\\" role=\\"img\\"><path d=\\"M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z\\"></path></svg></button><header class=\\"pf-c-modal-box__header\\"><h1 id=\\"pf-modal-part-1\\" class=\\"pf-c-modal-box__title\\"><span class=\\"pf-c-modal-box__title-text\\">Test Enable</span></h1></header><div id=\\"pf-modal-part-2\\" class=\\"pf-c-modal-box__body\\"><form novalidate=\\"\\" class=\\"pf-c-form\\"><div class=\\"pf-c-form__group\\"><div class=\\"pf-c-form__group-label\\"><label class=\\"pf-c-form__label\\" for=\\"Enter a Key\\"><span class=\\"pf-c-form__label-text\\">Enter a Key</span></label> </div><div class=\\"pf-c-form__group-control\\"><input id=\\"Enter a Key\\" class=\\"pf-c-form-control odh-enable-modal__variable-input\\" type=\\"password\\" aria-invalid=\\"false\\" value=\\"\\"><button aria-disabled=\\"false\\" class=\\"pf-c-button pf-m-link odh-enable-modal__toggle-password-vis\\" type=\\"button\\" data-ouia-component-type=\\"PF4/Button\\" data-ouia-safe=\\"true\\" data-ouia-component-id=\\"OUIA-Generated-Button-link-1\\"><svg style=\\"vertical-align: -0.125em;\\" fill=\\"currentColor\\" height=\\"1em\\" width=\\"1em\\" viewBox=\\"0 0 576 512\\" aria-hidden=\\"true\\" role=\\"img\\"><path d=\\"M572.52 241.4C518.29 135.59 410.93 64 288 64S57.68 135.64 3.48 241.41a32.35 32.35 0 0 0 0 29.19C57.71 376.41 165.07 448 288 448s230.32-71.64 284.52-177.41a32.35 32.35 0 0 0 0-29.19zM288 400a144 144 0 1 1 144-144 143.93 143.93 0 0 1-144 144zm0-240a95.31 95.31 0 0 0-25.31 3.79 47.85 47.85 0 0 1-66.9 66.9A95.78 95.78 0 1 0 288 160z\\"></path></svg></button><div class=\\"pf-c-form__helper-text\\" id=\\"Enter a Key-helper\\" aria-live=\\"polite\\">This key is enables the app</div></div></div></form></div><footer class=\\"pf-c-modal-box__footer\\"><button aria-disabled=\\"false\\" class=\\"pf-c-button pf-m-primary\\" type=\\"button\\" data-ouia-component-type=\\"PF4/Button\\" data-ouia-safe=\\"true\\" data-ouia-component-id=\\"OUIA-Generated-Button-primary-1\\">Enable</button><button aria-disabled=\\"false\\" class=\\"pf-c-button pf-m-link\\" type=\\"button\\" data-ouia-component-type=\\"PF4/Button\\" data-ouia-safe=\\"true\\" data-ouia-component-id=\\"OUIA-Generated-Button-link-2\\">Cancel</button></footer></div>"`;

exports[`ExploreApplications should show the getting started panel on card click 1`] = `
"<section class=\\"pf-c-page__main-section pf-m-light odh-apps__heading\\"><div class=\\"pf-c-content odh-apps__heading__text\\"><h1 data-pf-content=\\"true\\" class=\\"\\">Explore</h1><p data-pf-content=\\"true\\" class=\\"\\">Add optional applications to your Open Data Hub instance.</p></div></section><div class=\\"pf-c-drawer pf-m-expanded pf-m-inline\\"><div class=\\"pf-c-drawer__main\\"><div class=\\"pf-c-drawer__content\\"><div class=\\"pf-c-drawer__body odh-explore-apps__body m-side-panel-open\\"><div class=\\"odh-dashboard__page-content\\"><section class=\\"pf-c-page__main-section\\"><div class=\\"pf-l-gallery pf-m-gutter odh-explore-apps__gallery\\"><article id=\\"jupyterhub\\" class=\\"pf-c-card pf-m-hoverable pf-m-selectable odh-card\\" tabindex=\\"0\\" data-ouia-component-type=\\"PF4/Card\\" data-ouia-safe=\\"true\\" data-ouia-component-id=\\"OUIA-Generated-Card-3\\"><div class=\\"pf-c-card__header\\"><img class=\\"pf-c-brand odh-card__header-brand odh-card__header-brand\\" src=\\"images/jupyterhub.svg\\" alt=\\"JupyterHub\\"></div><div class=\\"pf-c-card__title\\"><span class=\\"odh-card__title-supported\\">JupyterHub</span><div><span class=\\"odh-card__provider\\">by Jupyter</span></div></div><div class=\\"pf-c-card__body\\">A multi-user version of the notebook designed for companies, classrooms and research labs.</div></article><article id=\\"test-app\\" class=\\"pf-c-card pf-m-hoverable pf-m-selectable odh-card\\" tabindex=\\"0\\" data-ouia-component-type=\\"PF4/Card\\" data-ouia-safe=\\"true\\" data-ouia-component-id=\\"OUIA-Generated-Card-4\\"><div class=\\"pf-c-card__header\\"><img class=\\"pf-c-brand odh-card__header-brand odh-card__header-brand\\" src=\\"images/jupyterhub.svg\\" alt=\\"Test App\\"></div><div class=\\"pf-c-card__title\\"><span class=\\"odh-card__title-supported\\">Test App</span><div><span class=\\"odh-card__provider\\">by Test</span></div></div><div class=\\"pf-c-card__body\\">Some description</div></article></div></section></div></div></div><div class=\\"pf-c-drawer__panel pf-m-resizable odh-get-started\\" style=\\"--pf-c-drawer__panel--md--FlexBasis--min: 350px;\\"><div class=\\"pf-c-drawer__splitter pf-m-vertical\\" role=\\"separator\\" tabindex=\\"0\\" aria-orientation=\\"vertical\\" aria-label=\\"Resize\\" aria-describedby=\\"Press space to begin resizing, and use the arrow keys to grow or shrink the panel. Press enter or escape to finish resizing.\\"><div class=\\"pf-c-drawer__splitter-handle\\" aria-hidden=\\"true\\"></div></div><div class=\\"pf-c-drawer__panel-main\\"><div class=\\"pf-c-drawer__body\\"><div class=\\"pf-c-drawer__head\\"><div class=\\"odh-get-started__header\\"><h1 class=\\"pf-c-title pf-m-xl\\">JupyterHub</h1><div><span class=\\"odh-get-started__header__provider\\">by Jupyter</span></div></div><div class=\\"pf-c-drawer__actions\\"><div class=\\"pf-c-drawer__close\\"><button aria-disabled=\\"false\\" aria-label=\\"Close drawer panel\\" class=\\"pf-c-button pf-m-plain\\" type=\\"button\\" data-ouia-component-type=\\"PF4/Button\\" data-ouia-safe=\\"true\\" data-ouia-component-id=\\"OUIA-Generated-Button-plain-1\\"><svg style=\\"vertical-align: -0.125em;\\" fill=\\"currentColor\\" height=\\"1em\\" width=\\"1em\\" viewBox=\\"0 0 352 512\\" aria-hidden=\\"true\\" role=\\"img\\"><path d=\\"M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z\\"></path></svg></button></div></div></div></div><div class=\\"pf-c-drawer__body odh-get-started__button-panel\\"><a class=\\"pf-c-button pf-m-primary\\" href=\\"https://jupyterhub.readthedocs.io/en/stable/getting-started/index.html\\" target=\\"_blank\\" rel=\\"noopener noreferrer\\"><span class=\\"odh-get-started__get-started-text\\">Get started</span><svg style=\\"vertical-align: -0.125em;\\" fill=\\"currentColor\\" height=\\"1em\\" width=\\"1em\\" viewBox=\\"0 0 512 512\\" aria-hidden=\\"true\\" role=\\"img\\"><path d=\\"M432,320H400a16,16,0,0,0-16,16V448H64V128H208a16,16,0,0,0,16-16V80a16,16,0,0,0-16-16H48A48,48,0,0,0,0,112V464a48,48,0,0,0,48,48H400a48,48,0,0,0,48-48V336A16,16,0,0,0,432,320ZM488,0h-128c-21.37,0-32.05,25.91-17,41l35.73,35.73L135,320.37a24,24,0,0,0,0,34L157.67,377a24,24,0,0,0,34,0L435.28,133.32,471,169c15,15,41,4.5,41-17V24A24,24,0,0,0,488,0Z\\"></path></svg></a></div><div class=\\"pf-c-drawer__body odh-get-started__body\\"><div class=\\"odh-markdown-view\\"><h1>Quickstart</h1>
<h2>Prerequisites</h2>
<p>Before installing JupyterHub, you will need:</p>
<ul>
<li><p>a Linux/Unix based system</p></li>
<li><p><a target=\\"_blank\\" rel=\\"noopener noreferrer\\" href=\\"https://www.python.org/downloads/\\">Python</a> 3.5 or greater</p>
<p>An understanding
of using <a target=\\"_blank\\" rel=\\"noopener noreferrer\\" href=\\"https://pip.pypa.io/en/stable/\\"><code>pip</code></a> or
<a target=\\"_blank\\" rel=\\"noopener noreferrer\\" href=\\"https://conda.io/docs/get-started.html\\"><code>conda</code></a> for
installing Python packages is helpful.</p></li>
<li><p><a target=\\"_blank\\" rel=\\"noopener noreferrer\\" href=\\"https://www.npmjs.com/\\">nodejs/npm</a></p>
<p><a target=\\"_blank\\" rel=\\"noopener noreferrer\\" href=\\"https://docs.npmjs.com/getting-started/installing-node\\">Install nodejs/npm</a>,
using your operating system's package manager.</p></li>
<li><p>If you are using <strong><code>conda</code></strong>, the nodejs and npm dependencies will be installed for
you by conda.</p></li>
<li><p>If you are using <strong><code>pip</code></strong>, install a recent version of
<a target=\\"_blank\\" rel=\\"noopener noreferrer\\" href=\\"https://docs.npmjs.com/getting-started/installing-node\\">nodejs/npm</a>.
For example, install it on Linux (Debian/Ubuntu) using:</p>
<pre><code>sudo apt-get install npm nodejs-legacy
</code></pre>
<p>The <code>nodejs-legacy</code> package installs the <code>node</code> executable and is currently
required for npm to work on Debian/Ubuntu.</p></li>
<li><p>A <a target=\\"_blank\\" rel=\\"noopener noreferrer\\" href=\\"https://en.wikipedia.org/wiki/Pluggable_authentication_module\\">pluggable authentication module (PAM)</a>
to use the <a target=\\"_blank\\" rel=\\"noopener noreferrer\\" href=\\"./getting-started/authenticators-users-basics.md\\">default Authenticator</a></p>
<p>PAM is often available by default on most distributions, if this is not the case it can be installed by
using the operating system's package manager.</p></li>
<li><p>TLS certificate and key for HTTPS communication</p></li>
<li><p>Domain name</p></li>
</ul>
<p>Before running the single-user notebook servers (which may be on the same
system as the Hub or not), you will need:</p>
<ul>
<li><a target=\\"_blank\\" rel=\\"noopener noreferrer\\" href=\\"https://jupyter.readthedocs.io/en/latest/install.html\\">Jupyter Notebook</a>
version 4 or greater</li>
</ul>
<h2>Installation</h2>
<p>JupyterHub can be installed with <code>pip</code> (and the proxy with <code>npm</code>) or <code>conda</code>:</p>
<p><strong>pip, npm:</strong></p>
<pre><code class=\\"bash language-bash\\">python3 -m pip install jupyterhub
npm install -g configurable-http-proxy
python3 -m pip install notebook  # needed if running the notebook servers locally
</code></pre>
<p><strong>conda</strong> (one command installs jupyterhub and proxy):</p>
<pre><code class=\\"bash language-bash\\">conda install -c conda-forge jupyterhub  # installs jupyterhub and proxy
conda install notebook  # needed if running the notebook servers locally
</code></pre>
<p>Test your installation. If installed, these commands should return the packages'
help contents:</p>
<pre><code class=\\"bash language-bash\\">jupyterhub -h
configurable-http-proxy -h
</code></pre>
<h2>Start the Hub server</h2>
<p>To start the Hub server, run the command:</p>
<pre><code class=\\"bash language-bash\\">jupyterhub
</code></pre>
<p>Visit <code>https://localhost:8000</code> in your browser, and sign in with your unix
credentials.</p>
<p>To <strong>allow multiple users to sign in</strong> to the Hub server, you must start
<code>jupyterhub</code> as a <em>privileged user</em>, such as root:</p>
<pre><code class=\\"bash language-bash\\">sudo jupyterhub
</code></pre>
<p>The <a target=\\"_blank\\" rel=\\"noopener noreferrer\\" href=\\"https://github.com/jupyterhub/jupyterhub/wiki/Using-sudo-to-run-JupyterHub-without-root-privileges\\">wiki</a>
describes how to run the server as a <em>less privileged user</em>. This requires
additional configuration of the system.</p></div></div></div></div></div></div>"
`;
